File: DSL Practical/Practical-6.py
def insertion(marks,n):
    for i in range(0,len(marks)):
        temp=marks[i]
        j=i-1
        while(j>=0 and marks[j]>temp):
            marks[j+1]=marks[j]
            j=j-1
        marks[j+1]=temp
    print(f"SORTED MARKS: {marks}")

def shell(marks):
    gap=len(marks)//2
    while (gap>0):
        for i in range(gap,len(marks)):
            temp=marks[i]
            j=i
            while(j>=gap and marks[j-gap]>temp):
                marks[j]=marks[j-gap]
                j=j-gap
            marks[j]=temp
        gap=gap//2
    print(f"SORTED MARKS: {marks}")

def top5(marks,n):
    print(marks[:n-66
                :-1])

def main():
    marks=[]
    num=int(input("ENTER THE NUMBER OF STUDENTS: "))
    for i in range(0,num):
        no=int(input("ENTER THE NUMBER: "))
        marks.append(no)
    while True:
        print("**************MENU**************\n1.Insertion sort\n2.Shell Sort\n3.TOP 5\n4.EXIT")
        choice=int(input("ENTER THE CHOICE: "))
        if choice==1:
            insertion(marks,len(marks))
        elif choice==2:
            shell(marks)
        elif choice==3:
            top5(marks,len(marks))
        elif choice==4:
            print("EXIT")
            break
        else:
            print("INVALID INPUT!!!!!")
main()

File: OOP Practical/Practical-6.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
class student
{
    public:
    string name,dob,tele;
    int roll,age;
    student()
    {
        name="";
        dob="";
        tele="";
        roll=0;
        age=0;
    }

    void getdata()
    {
        cin.ignore();
        cout<<"ENTER THE NAME OF STUDENT: "<<endl;
        getline(cin,name);
        cout<<"ENTER THE AGE OF STUDENT: "<<endl;
        cin>>age;
        cout<<"ENTER THE DATE OF BIRTH OF STUDENT:(DD/MM/YYYY) "<<endl;
        cin>>dob;
        cout<<"ENTER THE TELEPHONE NUMBER OF STUDENT: "<<endl;
        cin>>tele;
        cout<<"ENTER THE ROLL NUMBER OF STUDENT: "<<endl;
        cin>>roll;
    }

    void display()
    {
        cout<<"\nNAME: "<<name<<endl;
        cout<<"ROLL NUMBER: "<<roll<<endl;
        cout<<"DATE OF BIRTH: "<<dob<<endl;
        cout<<"AGE: "<<age<<endl;
        cout<<"TELEPHONE NUMBER: \n"<<tele<<endl;
    }

    bool operator ==(const student& obj)
    {
        return roll==obj.roll;
    }
};

bool compare(student& s1,student& s2)
{
    return s1.roll<s2.roll;
}

int main()
{
    int choice,num;
    vector<student>v1;
    do{
        cout<<"\n***************MENU***************\n1.ENTER DATA\n2.DISPLAY DATA\n3.SORTING DATA\n4.SEARCHING DATA\n5.EXIT"<<endl;
        cout<<"ENTER YOUR CHOICE: "<<endl;
        cin>>choice;
        switch(choice)
        {
            case 1:
            {
                cout<<"ENTER THE NUMBER OF STUDENTS: "<<endl;
                cin>>num;
                cout<<"ENTER THE DATA OF STUDENT: "<<endl;
                for(int i=0;i<num;i++)
                {
                    student s;
                    s.getdata();
                    v1.push_back(s);
                }
                cout<<"\nDATA ADDED TO THE VECTOR!\n"<<endl;
                break;
            }
            case 2:
            {
                cout<<"\nAll the entries are as follows: \n"<<endl;
                for(int i=0;i<v1.size();i++)
                {
                    v1[i].display();
                }
                cout<<"\nTotal data in the vector is \n"<<v1.size()<<endl;
                break;
            }
            case 3:
            {
                student s;
                sort(v1.begin(),v1.end(),compare);
                cout<<"data sorted successfully!!!"<<endl;
                break;
            }
            case 4:
            {
                student s;
                cout<<"Enter the roll of the Student you want to find: "<<endl;
                cin>>s.roll;
                vector<student>::iterator it;
                it=find(v1.begin(),v1.end(),s);
                if(it==v1.end())
                    cout<<"Student with roll number "<<s.roll<<" not found in Database!!"<<endl;
                else{
                    cout<<"STUDENT FOUND: "<<endl;
                    it->display();
                }
                break;
            }
            case 5:
            {
                cout<<"EXITING THE PROGRAM.....................:)"<<endl;
                break;
            }
            default:
            {
                cout<<"INVALID INPUT!!!"<<endl;
                break;
            }
        }
    }while(choice!=5);
    return 0;
}